File: backend/app/schemas/commute.py
"""Commute lookup request/response (TRN-01)."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

TransportMode = Literal["walk", "subway", "bus", "taxi", "flight", "ferry"]
RouteSource = Literal["directions", "user_hint", "estimate"]


class CommutePoint(BaseModel):
    lat: float = Field(..., ge=-90, le=90)
    lng: float = Field(..., ge=-180, le=180)
    name: str = ""


class CommuteLookupRequest(BaseModel):
    from_point: CommutePoint
    to_point: CommutePoint
    """User original prompt / notes — used for ferry/trail/cable-car hints."""
    free_text: str = ""
    notes: str = ""
    """When false, only hints + estimate (no SerpApi)."""
    use_directions: bool = True


class CommuteCandidate(BaseModel):
    transport_mode: TransportMode
    duration_minutes: int = Field(..., ge=0)
    distance_meters: int | None = None
    label: str = ""
    summary: str = ""
    source: RouteSource
    recommended: bool = False
    fare_estimate: float | None = None
    currency: str | None = None


class CommuteLookupResponse(BaseModel):
    candidates: list[CommuteCandidate] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)
    provider: str | None = None
    straight_line_meters: int | None = None

File: frontend/src/api/commute.ts
import type { RouteSource, TransportMode } from '../types/itinerary';

export interface CommutePoint {
  lat: number;
  lng: number;
  name?: string;
}

export interface CommuteCandidate {
  transport_mode: TransportMode;
  duration_minutes: number;
  distance_meters?: number | null;
  label: string;
  summary: string;
  source: RouteSource;
  recommended: boolean;
  fare_estimate?: number | null;
  currency?: string | null;
}

export interface CommuteLookupResponse {
  candidates: CommuteCandidate[];
  warnings: string[];
  provider?: string | null;
  straight_line_meters?: number | null;
}

export class CommuteApiError extends Error {
  status: number;

  constructor(message: string, status: number) {
    super(message);
    this.name = 'CommuteApiError';
    this.status = status;
  }
}

const apiBase = () => import.meta.env.VITE_API_BASE ?? '';

export async function lookupCommute(body: {
  from_point: CommutePoint;
  to_point: CommutePoint;
  free_text?: string;
  notes?: string;
  use_directions?: boolean;
}): Promise<CommuteLookupResponse> {
  const res = await fetch(`${apiBase()}/api/v1/commute/lookup`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      from_point: body.from_point,
      to_point: body.to_point,
      free_text: body.free_text ?? '',
      notes: body.notes ?? '',
      use_directions: body.use_directions ?? true,
    }),
  });
  if (!res.ok) {
    let detail = '通勤查询失败';
    try {
      const data = (await res.json()) as { detail?: string };
      if (data.detail) detail = data.detail;
    } catch {
      /* ignore */
    }
    throw new CommuteApiError(detail, res.status);
  }
  return (await res.json()) as CommuteLookupResponse;
}
